c_src/sasl_auth.c

```diff
@@ -1,291 +1,355 @@
 #include <erl_nif.h>
-#include <stdio.h>
-#include <string.h>
-#include <sasl/sasl.h>
 #include <krb5.h>
+#include <sasl/sasl.h>
+#include <string.h>
 
 // MIT is broken
 #ifdef KRB5_KRB5_H_INCLUDED
 #define KRB5KDC_ERR_KEY_EXPIRED KRB5KDC_ERR_KEY_EXP
 #endif
 
-// Structure for kerberos authentication
-struct kebab
-{
-  krb5_error_code error;
-  krb5_principal  principal;
-  krb5_context    context;
-  krb5_creds      creds;
-};
-
-typedef struct sasl_state_s {
-        sasl_conn_t *conn;
-        sasl_callback_t callbacks[16];
+#ifdef UNUSED
+#elif defined(__GNUC__)
+#define UNUSED(x) UNUSED_##x __attribute__((unused))
+#elif defined(__LCLINT__)
+#define UNUSED(x) /*@unused@*/ x
+#else
+#define UNUSED(x) x
+#endif
+
+ERL_NIF_TERM ATOM_OK;
+ERL_NIF_TERM ATOM_ERROR;
+ERL_NIF_TERM ATOM_OOM;
+ERL_NIF_TERM ATOM_UNKNOWN;
+
+#define OK_TUPLE(env, ret) enif_make_tuple2(env, ATOM_OK, ret)
+#define SASL_STEP_TUPLE(env, code, out, outlen)                                                    \
+    OK_TUPLE(env, enif_make_tuple2(env, enif_make_int(env, code), str_to_bin(env, out, outlen)))
+#define ERROR_TUPLE(env, ret) enif_make_tuple2(env, ATOM_ERROR, ret)
+#define SASL_ERROR_TUPLE(env, state, code)                                                         \
+    ERROR_TUPLE(env, enif_make_tuple2(env, enif_make_int(env, code), sasl_error(env, state)));
+
+typedef struct {
+    sasl_conn_t* conn;
+    sasl_callback_t callbacks[16];
+    int mech_set;
 } sasl_state_t;
 
 // SASL connection state
-sasl_state_t *state;
+static ErlNifResourceType* sasl_resource = NULL;
```

Comment:
Maybe we should put this in priv_data (see https://www.erlang.org/doc/man/erl_nif.html#enif_priv_data) to make upgrading the library easier. However, in that case I think we can do that in another PR.

Reply:
Have seen this done, I'd say a subsequent PR makes sense to me, there's some other minor changes I'd like to get in along side that as well.